h5pt.go

```diff
@@ -82,170 +90,154 @@ func (t *Table) ReadPackets(start, nrecords int, data interface{}) error {
+		val := C.int8_t(int8(rv.Int()))
+		dataPtr = unsafe.Pointer(&val)
+		dataSize = 1
 
 	case reflect.Uint8:
-		val := C.uint8_t(data.(uint8))
-		cData = unsafe.Pointer(&val)
-
-	case reflect.Uint16:
-		val := C.uint16_t(data.(uint16))
-		cData = unsafe.Pointer(&val)
+		ad.noRecords = 1
+		val := C.uint8_t(uint8(rv.Uint()))
+		dataPtr = unsafe.Pointer(&val)
+		dataSize = 1
 
 	case reflect.Int16:
-		val := C.int16_t(data.(int16))
-		cData = unsafe.Pointer(&val)
+		ad.noRecords = 1
+		val := C.uint8_t(int16(rv.Int()))
+		dataPtr = unsafe.Pointer(&val)
+		dataSize = 2
+
+	case reflect.Uint16:
+		ad.noRecords = 1
+		val := C.uint8_t(uint16(rv.Uint()))
+		dataPtr = unsafe.Pointer(&val)
+		dataSize = 2
 
 	case reflect.Int32:
-		val := C.int32_t(data.(int32))
-		cData = unsafe.Pointer(&val)
+		ad.noRecords = 1
+		val := C.uint8_t(int32(rv.Int()))
+		dataPtr = unsafe.Pointer(&val)
+		dataSize = 4
 
 	case reflect.Uint32:
-		val := C.uint32_t(data.(uint32))
-		cData = unsafe.Pointer(&val)
+		ad.noRecords = 1
+		val := C.uint8_t(uint32(rv.Uint()))
+		dataPtr = unsafe.Pointer(&val)
+		dataSize = 4
 
 	case reflect.Int64:
-		val := C.int64_t(data.(int64))
-		cData = unsafe.Pointer(&val)
+		ad.noRecords = 1
+		val := C.int64_t(int64(rv.Int()))
+		dataPtr = unsafe.Pointer(&val)
+		dataSize = 8
 
 	case reflect.Uint64:
-		val := C.uint64_t(data.(uint64))
-		cData = unsafe.Pointer(&val)
+		ad.noRecords = 1
+		val := C.uint64_t(uint64(rv.Uint()))
+		dataPtr = unsafe.Pointer(&val)
+		dataSize = 8
 
 	case reflect.Float32:
-		val := C.float(data.(float32))
-		cData = unsafe.Pointer(&val)
+		ad.noRecords = 1
+		val := C.float(float32(rv.Float()))
+		dataPtr = unsafe.Pointer(&val)
+		dataSize = 4
 
 	case reflect.Float64:
-		val := C.double(data.(float64))
-		cData = unsafe.Pointer(&val)
+		ad.noRecords = 1
+		val := C.double(float64(rv.Float()))
+		dataPtr = unsafe.Pointer(&val)
+		dataSize = 8
+
+	case reflect.Bool:
+		ad.noRecords = 1
+		val := C.uchar(0)
+		if rv.Bool() {
+			val = 1
+		}
+		dataPtr = unsafe.Pointer(&val)
+		dataSize = 1
 
 	default:
-		return fmt.Errorf("hdf5: PT Append does not support datatype (%s).", rt.Kind())
+		err = fmt.Errorf("hdf5: PT Append does not support datatype (%s).", rt.Kind())
+		return
+	}
+
+	ad.memSize = dataSize + ad.offset
+	ad.ptr = C.realloc(ad.ptr, ad.memSize)
+	C.memcpy(unsafe.Pointer(uintptr(ad.ptr)+uintptr(ad.offset)), dataPtr, dataSize)
+	ad.offset += dataSize
+
+	return
+}
+
+// Append appends packets to the end of a packet table.
+func (t *Table) Append(data interface{}) (err error) {
+	ad := appendData{
+		ptr: unsafe.Pointer(nil),
+		// Ensure to start with zero initialization on curcial fields
+		memSize: C.size_t(0),
+		offset:  C.size_t(0),
+	}
+	defer C.free(ad.ptr)
+
+	if err = extractValues(&ad, data); err != nil {
+		return
+	}
+
+	// Make sure that we free all memory that was allocated to strings after
+	// they were written to the HDF5 file.
+	for _, str := range ad.strs {
+		defer C.free(str)
```

Comment:
instead of `len(ad.strs)` `defer`s, couldn't we create only one `defer` whose body does the looping + freeing?

Reply:
actually, we could refactor all these `defer`s into one method on `appendData`:
```go
func (ad *appendData) free() {
    C.free(ad.ptr)
    for _, str := range ad.strs {
        C.free(str)
    }
}
```

so we could then just write:
```go
var ad appendData
defer ad.free()
```

Reply:
Nice! Will do that.